rfc/tap-hardening.md

```diff
@@ -0,0 +1,125 @@
+# Summary
+
+The goal of the Tap Hardening project is to restrict the set of users by
+integrating with native Kubernetes RBAC. The current proposal is for the Tap
+Service to become an [Extension API server](https://kubernetes.io/docs/tasks/access-kubernetes-api/setup-extension-api-server/). This will allow requests to be
+both authorized and authenticated through the Kubernetes API. The TapByResource
+route currently served by the public API will be served by the APIService object
+on the now separate Tap Pod.
+
+The data plane proxies will also begin authenticating and authorizing Tap
+requests. The proxy’s Tap server will require a TLS connection and the client
+must be the Tap Controller. All other authenticated clients will be rejected.
+
+# Motivation
+
+The Tap feature is open to any user that has access to the cluster. Tap requests
+made through the CLI and the Web Deployment are neither authenticated or
+authorized. The TapByResource route is served by the public API and will start
+Tap connections to any of the resources requested.
+
+When the Linkerd proxies for requested resources receive tap requests, the Tap
+server also does not authenticate or authorize the requesting client. If it can
+open a connection over TLS it will, but it is not required.
+
+# Details
+
+## Control Plane
+
+The Tap Controller becoming an Extension API server will allow requests to be
+authenticated and authorized through an [aggregation layer](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/). The Tap APIService
+will claim the TapByResource route in the Kubernetes API. This will allow the
+aggregation layer to proxy anything sent to that API path to the Tap extension
```

Comment:
The creation of the APIService resource should probably follow the same pattern as the webhooks use which is described in https://github.com/linkerd/linkerd2/pull/2806

i.e. The cert, key, and APIService resource are generated at install time and the cert and key are loaded by the tap controller when it starts up.